### PR TITLE
Implemented search by title and author

### PR DESCRIPTION
**Pull request by Trayvor**

### CODE REVIEW

File: src/main/java/org/example/bookstore/exception/KeyNotFoundException.java [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package org.example.bookstore.exception;
+
+public class KeyNotFoundException extends RuntimeException {
+    public KeyNotFoundException(String message, Throwable throwable) {
+        super(message, throwable);
```
Comment (okuzan) [inline src/main/java/org/example/bookstore/exception/KeyNotFoundException.java]: good approach!

File: src/main/java/org/example/bookstore/repository/book/BookSpecificationProviderManager.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+@RequiredArgsConstructor
+@Component
+public class BookSpecificationProviderManager implements SpecificationProviderManager<Book> {
+    private final List<SpecificationProvider<Book>> bookSpecificationProviders;
+
+    @Override
+    public SpecificationProvider<Book> getSpecificationProvider(String key) {
+        return bookSpecificationProviders.stream()
+                .filter(p -> p.getKey().equals(key))
```
Comment (okuzan) [inline src/main/java/org/example/bookstore/repository/book/BookSpecificationProviderManager.java]: change `p` to `provider`

File: src/main/java/org/example/bookstore/repository/book/spec/AuthorSpecificationProvider.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+    @Override
+    public String getKey() {
+        return "author";
+    }
+
+    @Override
+    public Specification<Book> getSpecification(String param) {
+        String pattern = "%" + param + "%";
+        return (root, query, criteriaBuilder) ->
+                criteriaBuilder.like(criteriaBuilder.lower(root.get("author")),
```
Comment (okuzan) [inline src/main/java/org/example/bookstore/repository/book/spec/AuthorSpecificationProvider.java]: let's extract repeated literal to constant

File: src/main/java/org/example/bookstore/repository/book/spec/TitleSpecificationProvider.java [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+    @Override
+    public String getKey() {
+        return "title";
+    }
+
+    @Override
+    public Specification<Book> getSpecification(String param) {
+        String pattern = "%" + param + "%";
+        return (root, query, criteriaBuilder) ->
+                criteriaBuilder.like(criteriaBuilder.lower(root.get("title")),
```
Comment (okuzan) [inline src/main/java/org/example/bookstore/repository/book/spec/TitleSpecificationProvider.java]: same here

**Review by okuzan — CHANGES_REQUESTED** [review]

**Review by sarakhmen — APPROVED** [review]

**Pull request merged**